packages/astro/src/core/build/generate.ts

```diff
@@ -156,16 +171,17 @@ async function generatePage(
 	// Get paths for the route, calling getStaticPaths if needed.
 	const paths = await getPathsForRoute(pageData, pageModule, opts, builtPaths);
 
-	for (let i = 0; i < paths.length; i++) {
-		const path = paths[i];
-		await generatePath(path, opts, generationOptions);
-		const timeEnd = performance.now();
-		const timeChange = getTimeStat(timeStart, timeEnd);
-		const timeIncrease = `(+${timeChange})`;
-		const filePath = getOutputFilename(opts.astroConfig, path, pageData.route.type);
-		const lineIcon = i === paths.length - 1 ? '└─' : '├─';
-		info(opts.logging, null, `  ${cyan(lineIcon)} ${dim(filePath)} ${dim(timeIncrease)}`);
+	const concurrency = opts.buildConfig.concurrency || MAX_CONCURRENT_RENDERS;
+	if (concurrency > 1) {
+		info(opts.logging, null, `${magenta('⚠️')}  Building with ${concurrency} concurrent renders`);
 	}
+	// Throttle the paths to avoid overloading the CPU with too many tasks.
+	for await (let _ of generatePathsConcurrently(concurrency, opts, generationOptions, timeStart, paths.values(), pageData.route.type)) {
```

Comment:
So now that I understand why generatePathsConcurrently is doing I can see that we are only getting max 5 at a time.

However, given that this code is inside of generatePage doesn't that mean that you aren't actually always generating 5 paths at a time?

If you are generating a route with many paths, such as when using getStaticPaths you will be. But if you have, for example, 5000 markdown files in `src/pages` you would still only be doing them 1 at a time, because each is a different route/page.


So would it be better to move the concurrency logic up a level so that you always get 5 concurrent renders?

Reply:
Maybe, but I think the getStaticPaths case is more common for a lot of sites.

The problem is the way the rendering code is currently this would be harder to implement without rewriting more of the logic. 
If someone has this use case I suppose they can look at it, I don't have it personally so I wouldn't be able to test it.

Reply:
We definitely see projects with a lot of individual .md files.

Reply:
Yeah that won't change anything for them, but it's probably a good start if someone wants to change the rendering code later.